dsl_lang/lib/util.ml

```diff
@@ -1,44 +1,45 @@
 exception NonTerminal
 exception ParserError of string
 
-(** syntax_error_msg lexbuf is a syntax error message for the current position **)
-let syntax_error_msg lexbuf =
-  let pos = Lexing.lexeme_start_p lexbuf in
-  let lnum, cnum = (pos.pos_lnum, pos.pos_cnum - pos.pos_bol) in
-  Printf.sprintf "Syntax error at line %d, character %d" lnum cnum
-
-let parse lexbuf = Parser.prog Lexer.token lexbuf
-
-(** parse s parses a program string into an AST **)
-let parse_string (s : string) =
-  let lexbuf = Lexing.from_string s in
-  try parse lexbuf
-  with Parser.Error -> raise (ParserError (syntax_error_msg lexbuf))
-
-(** parse s parses a program file into an AST **)
-let parse_file (f : string) =
-  let lexbuf = Lexing.from_channel (open_in f) in
-  try parse lexbuf with
-  | Parser.Error ->
-      prerr_endline (syntax_error_msg lexbuf);
-      exit 1
-  | Lexer.Err ->
-      prerr_endline (syntax_error_msg lexbuf);
-      exit 1
-
-(** Helper function to turn policy lists into strings. **)
-let rec string_of_plist (plist : Ast.policy list) : string =
-  match plist with
-  | [] -> ""
-  | [ x ] -> string_of_policy x
-  | h :: t -> string_of_policy h ^ ", " ^ string_of_plist t
-
 (** Takes a policy and returns the string representation of it. **)
-and string_of_policy (pol : Ast.policy) : string =
+let rec string_of_policy (pol : Ast.policy) : string =
+  
+  (* Helper function to compactly join policy lists by comma *)
+  let join = List.fold_left (fun acc s ->
+    let parsed = string_of_policy s in
+    if acc = "" then parsed
+    else (acc ^ ", " ^ parsed)) "" in
+
+  (* Helper function to compactly join weighted policy lists by comma *)
+  let join_weighted = List.fold_left (fun acc s ->
+    let parsed = "(" ^ string_of_policy (fst s) ^ ", " ^ string_of_int (snd s) ^ ")" in
+    if acc = "" then parsed
+    else (acc ^ ", " ^ parsed)) "" in
+
```

Comment:
I noticed you and @csziklai would manually do the list concatenation yourselves. There's a library function that does this for you: `String.concat`. For example, you could write `join` like this:
```
let join lst = lst |> List.map string_of_policy |> String.concat ", "
```

This is super insignificant though; I was just sharing some OCaml trivia

Reply:
Thanks! Perfect and I found a way to use this idea for `join_weighted` as well.